BUG: preserve timezone info when writing empty tz-aware series to HDF5

arw2019 opened the pull request:
- [x] closes #20594 (coupled with #37069 for empty tz-aware dataframes)
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

xref #20595 (stale PR that looked into this)

arw2019 committed:
TST: HDF5 roundtrip of tz-aware series (empty & non-empty)

arw2019 committed:
BUG/REF: rewrite write/write_array to handle tz info for series

arw2019 committed:
update TODO

arw2019 committed:
revert merge errors

arw2019 committed:
apply patch from #37069

jreback commented on pandas/io/pytables.py:
```diff
@@ -2960,11 +2960,17 @@ def write_array_empty(self, key: str, value: ArrayLike):
 
-    def write_array(self, key: str, value: ArrayLike, items: Optional[Index] = None):
-        # TODO: we only have one test that gets here, the only EA
+    def write_array(self, key: str, obj, items: Optional[Index] = None):
```
why is the typing changed? can you be explicit here

arw2019 replied:
Added a type hint.

AFAIU for series `to_numpy()` drops timezone info so we need to pass the original object in `SeriesFixed.read(...)`and special case here on timezone-aware data

xref (from stale PR that worked on this) https://github.com/pandas-dev/pandas/pull/20595/files#r179816447

jreback commented on pandas/io/pytables.py:
```diff
@@ -2960,11 +2960,17 @@ def write_array_empty(self, key: str, value: ArrayLike):
+        elif is_datetime64tz_dtype(obj):
+            value = obj.dt._get_values()
+        else:
+            value = obj.values
```
use .to_numpy()

arw2019 replied:
Done

jreback reviewed (changes requested):
ok i am not sure how to grok this one as i think its on top of the other;  in the future put them both in 1 PR is ok for this size change, or now, wait till the other is merged then rebase this one (or make this orthogonal)

arw2019 commented:
> ok i am not sure how to grok this one as i think its on top of the other; in the future put them both in 1 PR is ok for this size change, or now, wait till the other is merged then rebase this one (or make this orthogonal)

Ok will do! For this one I'll wait until the other is merge (since this is a dependant)

arw2019 committed:
use to_numpy() instead of values

arw2019 committed:
in write_array add type hint to obj

arw2019 committed:
rewrite tests using tz_aware_fixture

arw2019 committed:
DOC: add whatsnew

arw2019 commented:
@jreback this is ready for re-review
- [x] rebased against master (includes  #37069)
- [x] rewrote the tests from here and  #37069 using `tz_aware_fixture`
- [x] added whatsnew

arw2019 committed:
Merge remote-tracking branch 'upstream/master' into GH20594

arw2019 committed:
Merge remote-tracking branch 'upstream/master' into GH20594

arw2019 committed:
Merge remote-tracking branch 'upstream/master' into GH20594

jbrockmendel commented on pandas/io/pytables.py:
```diff
@@ -2968,11 +2968,17 @@ def write_array_empty(self, key: str, value: ArrayLike):
         #  both self._filters and EA
-        assert isinstance(value, (np.ndarray, ABCExtensionArray)), type(value)
+
+        if isinstance(obj, (np.ndarray, ABCExtensionArray)):
```
is the annotation of obj as a Series accurate?  if so, this branch is never hit

arw2019 replied:
No it's not. Should be SeriesOrFrame (will fix)

jbrockmendel commented on pandas/io/pytables.py:
```diff
@@ -2968,11 +2968,17 @@ def write_array_empty(self, key: str, value: ArrayLike):
+        if isinstance(obj, (np.ndarray, ABCExtensionArray)):
+            value = obj
+        elif is_datetime64tz_dtype(obj):
+            value = obj.dt._get_values()
```
i dont know off the top of my head what _get_values returns, is it a DatetimeArray?  The standard pattern would be to use `extract_array(obj, extract_numpy=True)` at the beginning of this method

arw2019 replied:
`_get_values` returns a `DatetimeIndex`

https://github.com/pandas-dev/pandas/blob/10e5ad7109f1bb3c47b48b6f0df185e108c5493d/pandas/core/indexes/accessors.py#L47-54

But actually the `extract_array` pattern does the job of this entire if/elif/else block here - thanks for suggesting! (I'll remember for future)

arw2019 committed:
Merge remote-tracking branch 'upstream/master' into GH20594

arw2019 committed:
feedback: use extract_array instead of special casing

arw2019 committed:
TYP/BUG: fix type hint in write_array signature

arw2019 committed:
Merge remote-tracking branch 'upstream/master' into GH20594

arw2019 committed:
Merge remote-tracking branch 'upstream/master' into GH20594

arw2019 committed:
merge with upstream

arw2019 commented:
CI Green + ready for re-review

jreback commented:
thanks @arw2019
